relayer/src/kill_switch.rs

```diff
@@ -0,0 +1,301 @@
+                // will trigger the emergency stop condition (i.e. the kill switch) in relayer contract.
+                // After that, there's no point in continuing because the relayer will be stopped/in emergency mode.
+                // Though, we need to wait for the kill switch transaction to be finalized.
+                self.state = State::WaitingForKillSwitchTxFin { tx_hash, proof };
+                return Ok(());
+            }
+        }
+
+        // After processing all events, `last_finalized_block` is the last block we've processed.
+        // So, we need to increment it by 1 to set the next block to process.
+        self.from_eth_block = Some(last_finalized_block.saturating_add(1));
+
+        Ok(())
+    }
+
+    async fn check_kill_switch_tx_finalized(&mut self, tx_hash: TxHash) -> anyhow::Result<()> {
+        log::info!("Checking for kill switch tx to be finalized");
+
+        let tx_status = self.eth_api.get_tx_status(tx_hash).await?;
+
+        match tx_status {
+            TxStatus::Finalized => {
+                log::info!("Kill switch tx finalized, exiting ..");
+                process::exit(0);
+            }
+            TxStatus::Pending => (),
+            TxStatus::Failed => {
+                log::warn!("Re-trying kill switch tx #{} finalization", tx_hash);
+
+                let State::WaitingForKillSwitchTxFin { tx_hash, proof } = &mut self.state else {
+                    unreachable!("Invalid state");
+                };
+
+                let new_tx_hash = submit_proof_to_ethereum(&self.eth_api, proof.clone()).await?;
+                // Update hash of the new kill switch transaction
+                *tx_hash = new_tx_hash;
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn get_block_finality_from_storage(
+        &mut self,
+        block_number: u64,
+    ) -> anyhow::Result<Option<BlockFinalityProofWithHash>> {
+        // NOTE: we use 32bits BE keys for block finality storage.
+        let key_bytes = (block_number as u32).to_be_bytes();
+
+        let value_bytes = match self.block_finality_storage.get(key_bytes)? {
+            Some(bytes) => bytes,
+            None => {
+                // Block finality proof generated for series of block,
+                // for example:
+                //  ```text
+                //          Round 1: finality proof for block #2 (transitively for 2,1 blocks)
+                //           |
+                //           |           Round 2: finality proof for block #5 (transitively for 5,4,3.. blocks)
+                //           |           |
+                //           v           v
+                //      [1] [2] [3] [4] [5] [6] [7] [8] [9] [10]
+                //  ```
+                // Some if we don't have the proof for block #3 or #4, we can try to get the proof for block #5.
+                let res = self.block_finality_storage.get_gt(key_bytes)?;
```

Comment:
Actually submitting some other block will not lead to a bridge halt, so I think `None` should be returned

Reply:
And, ofc, this error should be reported somehow at the `main_loop` as it means bridge's total failure

Reply:
Fixed, added metric.